File: internal/deeplink/deeplink.go
// Package deeplink parses and builds the gadak:// URLs that hand a piece of
// gadak to someone — a link in chat or on a web page instead of a shell, a
// running serve, and a command with side effects.
//
// # The grammar
//
//	gadak://<action>[/w/<profile>][/<subject>][?<params>]
//
//	gadak://view?pj=GDK&sc=inprogress        a filtered list, primary mirror
//	gadak://view/w/oss?issue=GDK-119         the same, on the "oss" mirror
//
// The action is the host. The optional `/w/<profile>` segment is the same one
// the web UI uses for a workspace mount, in the same position, so a gadak://
// link and the http:// link `gadak views open` prints describe the same thing
// the same way. The optional subject that may follow is what the action acts
// on — an issue key, a page key — and the query is passed through verbatim.
//
// # Why the whole grammar exists for one action
//
// Only `view` is implemented today, and this package deliberately does not
// know that: it validates the *shape* and returns whichever action it read.
// Deciding which actions exist belongs to the app that handles them
// (desktop/deeplink.go), not to the parser.
//
// The split matters because these two halves ship separately. A link lives in
// a chat log forever and is opened by whatever version happens to be
// installed, so the grammar is the part that must not change — while the set
// of actions grows every time a surface becomes addressable. Keeping the
// action list out of the parser means growing it is a table entry in the
// handler, and an older app meeting a newer link produces "this link needs a
// newer Gadak" instead of a parse error.
//
// What `view` can say is decided on the web side: the hash carries the view
// params (filters, display) and the place params (which panel and screen —
// issue, doc, person, feed, settings tab…). The registry of place params is
// web/src/lib/url-state.ts; a param registered there is linkable from here
// the same moment, with no change in this package. A second action is for
// what a hash cannot express — a place with no URL, or a different kind of
// address entirely.
//
// # Security posture
//
// Any web page can embed a gadak:// link, so the worst one may achieve is
// that the user briefly looks at the wrong thing. The grammar carries no verb
// and no payload beyond an address: a link says *where to go*, never what to
// do. Handlers must keep that true — an action that writes, or that submits
// anything, does not belong in this scheme however convenient it would be.
//
// Validation here is shape and size only:
//
//   - a well-formed action, an optional profile that is a safe directory
//     name, an optional subject that cannot escape a path,
//   - length and arity limits that bound what a handler is asked to parse.
//
// There is deliberately no allowlist of view parameter keys. The keys the UI
// understands are owned by VIEW_PARAM_KEYS in web/src/lib/view-config; a
// second copy here would drift the moment a new axis is added there, and the
// drift fails silently in the direction that matters — a link would stop
// working with no error anywhere. Unknown keys pass through and are ignored
// by the UI, which is the "wrong thing" worst case already accepted above.
package deeplink

import (
	"errors"
	"fmt"
	"net/url"
	"regexp"
	"strings"
)

// Callers must be able to tell these apart: a handler stays silent when the
// link simply is not ours (another app's scheme reaching a shared handler)
// and logs when it is ours but refused.
var (
	// ErrNotGadak reports a URL that is not a gadak:// link at all.
	ErrNotGadak = errors.New("deeplink: not a gadak:// URL")

	// ErrMalformed reports a gadak:// link this package refuses to parse.
	// Wrapping carries the specific rule that fired.
	ErrMalformed = errors.New("deeplink: malformed gadak:// URL")
)

// Scheme is the URL scheme, registered by the macOS bundle in
// desktop/build-app.sh. One owner, because the two are in a .go and a .sh
// file that nothing else connects and a disagreement is silent.
const Scheme = "gadak"

// ActionView shows a list of issues, filtered by the link's hash. The only
// action implemented today; see the package comment for why the parser does
// not enforce that.
const ActionView = "view"

// profileSegment is the path segment that introduces a profile, matching the
// web UI's /w/<name> workspace mount. Reserved: a subject may not be "w".
const profileSegment = "w"

// Size and shape limits. A deep link is typed or pasted by humans and handed
// over in chat, so real links are short; every constant here exists to bound
// what a handler can be asked to parse.
const (
	// maxInputLen bounds the whole link. URL handlers receive attacker-chosen
	// strings; 2048 bytes is far above any real view hash.
	maxInputLen = 2048

	// maxParams bounds how many parameters a link may carry. Real views use a
	// handful of axes (q, ks, pj, sc, a few f.<alias> filters); 32 leaves
	// headroom while capping the validation loop.
	maxParams = 32

	// maxValueLen bounds one parameter's value. Values are key lists and
	// search strings; 512 bytes is an order of magnitude above a real one.
	maxValueLen = 512

	// maxKeyLen bounds a parameter key. The static keys are two-letter
	// aliases, but a discovered-field axis is "f." + an ASCIISlug of a Jira
	// field's display name (internal/fields/slug.go), so its length is
	// whatever someone called a custom field. Measured across this machine's
	// profiles the longest real key is 19 bytes (f.requirement_11461), but
	// "Acceptance Criteria Verification Status" would slug to 41 — a
	// plausible field name whose link would silently do nothing. The cap
	// bounds the validation loop, and 64 bounds it as well as 32 does.
	maxKeyLen = 64
)

var (
	// actionPattern is the shape of an action. Lowercase words, hyphens
	// allowed for a future two-word action; no dots, so an action can never
	// be mistaken for a hostname.
	actionPattern = regexp.MustCompile(`^[a-z][a-z0-9-]{0,31}$`)

	// profilePattern mirrors config's profile-name rule: profile names become
	// directory names, so the first character is alphanumeric and the rest
	// contain no path separator or shell metacharacter.
	profilePattern = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9_.-]{0,63}$`)

	// subjectPattern is what an action may act on: an issue key (GDK-119), a
	// page key, an email. Permissive in character, strict in what it cannot
	// be — no separators, and the traversal names are refused outright below.
	subjectPattern = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9_.@+-]{0,127}$`)

	// keyPattern is the shape of a view-parameter key: the UI router's short
	// lowercase aliases (q, ks, pj, sc, f.<alias>). Shape only — keys are
	// never matched against the real key list; see the package comment. Built
	// from maxKeyLen so the limit has one owner and tests can derive it.
	keyPattern = regexp.MustCompile(fmt.Sprintf(`^[a-z][a-z0-9_.]{0,%d}$`, maxKeyLen-1))
)

// A Link is a parsed gadak:// URL.
type Link struct {
	// Action is what the link asks for, lowercased. Compare against the
	// Action* constants; an unrecognised one is a link from a newer gadak,
	// not an error.
	Action string

	// Profile is the mirror the link addresses. "" means the primary one;
	// "default" is accepted and means the same, because config does.
	Profile string

	// Subject is what the action acts on, or "" when it needs none. No action
	// uses it yet; the slot is reserved so adding one (gadak://issue/GDK-119)
	// does not have to change the grammar.
	Subject string

	// Hash is the view hash: the raw query string exactly as the UI's hash
	// router and uifocus want it, with no leading "#" or "?" and
	// percent-encoding untouched. Returned verbatim — never decoded,
	// re-encoded, or re-ordered — because the CLI's composeServeURL treats it
	// as opaque too, and any rewrite here would make the two paths disagree.
	// Empty when the link carries no query; whether that is acceptable is the
	// action's rule, not the grammar's.
	Hash string
}

// Parse turns a gadak:// URL into a Link.
//
// The host is compared case-insensitively and returned lowercased; the path
// is not, because profile and subject are identifiers. One trailing slash
// before the query is tolerated, because a human retyping a link produces
// one.
//
// err is ErrNotGadak when the URL is not a gadak:// link at all (the caller
// should stay silent) and wraps ErrMalformed when it is one this package
// rejects (the caller should say so).
func Parse(raw string) (Link, error) {
	u, err := url.Parse(raw)
	if err != nil {
		// Unparseable, but still ours if it claims our scheme prefix.
		if hasSchemePrefix(raw) {
			return Link{}, malformedf("unparseable URL: %v", err)
		}
		return Link{}, ErrNotGadak
	}
	if !strings.EqualFold(u.Scheme, Scheme) {
		return Link{}, ErrNotGadak
	}
	if len(raw) > maxInputLen {
		return Link{}, malformedf("input is %d bytes, over the %d-byte limit", len(raw), maxInputLen)
	}
	// The grammar carries no fragment. Checking the raw string also catches a
	// bare trailing "#", which url.Parse reports as an empty Fragment no
	// different from none (measured; see the tests).
	if strings.ContainsRune(raw, '#') {
		return Link{}, malformedf("fragment not allowed")
	}
	if u.User != nil {
		return Link{}, malformedf("userinfo not allowed")
	}
	// net/url preserves host case (measured) and keeps any port in Host, so
	// this fold is load-bearing and "view:7777" fails the pattern.
	action := strings.ToLower(u.Host)
	if !actionPattern.MatchString(action) {
		return Link{}, malformedf("action %q is not a plain lowercase word", u.Host)
	}
	profile, subject, err := splitPath(u.Path)
	if err != nil {
		return Link{}, err
	}
	if err := validateQuery(u.RawQuery); err != nil {
		return Link{}, err
	}
	return Link{Action: action, Profile: profile, Subject: subject, Hash: u.RawQuery}, nil
}

// Compose builds a gadak:// link: the inverse of Parse, in the package that
// owns the grammar, so the two cannot drift into disagreeing about the shape.
//
// prefix is the workspace segment — "" for the primary mirror, "/w/<name>"
// otherwise — and is taken as a parameter rather than derived here on
// purpose. The rule mapping a profile onto that segment belongs to
// internal/workspace, which pulls in the server and the sync engine; a parser
// any process might run should not drag those along to answer a question
// about a URL. Callers pass workspace.Prefix(profile, "").
//
// An empty hash yields "" for ActionView: there is no view to link to, and a
// link to nothing is worse than no link, because it looks like one that works.
func Compose(action, prefix, hash string) string {
	if action == ActionView && hash == "" {
		return ""
	}
	out := Scheme + "://" + action + prefix
	if hash != "" {
		out += "?" + hash
	}
	return out
}

// hasSchemePrefix reports whether raw begins with "gadak:", case-insensitively
// and without allocating — used to classify unparseable input.
func hasSchemePrefix(raw string) bool {
	const prefix = Scheme + ":"
	return len(raw) >= len(prefix) && strings.EqualFold(raw[:len(prefix)], prefix)
}

// splitPath reads the optional profile and subject out of the decoded path.
//
// Accepted: "", "/", "/w/<profile>", "/<subject>", "/w/<profile>/<subject>",
// each with an optional trailing slash. A leading "w" segment is reserved as
// the profile introducer, so a subject may not be the single letter "w" —
// no issue key, page key, or address is.
//
// The check runs on u.Path, the decoded form: an encoded separator (%2f) can
// only add segments, which the segment count and per-segment patterns then
// reject, so decoding cannot smuggle a separator into one identifier.
func splitPath(path string) (profile, subject string, err error) {
	trimmed := strings.Trim(path, "/")
	if trimmed == "" {
		return "", "", nil
	}
	seg := strings.Split(trimmed, "/")
	if seg[0] == profileSegment {
		if len(seg) < 2 {
			return "", "", malformedf("empty profile after /%s/", profileSegment)
		}
		if len(seg) > 3 {
			return "", "", malformedf("path %q has more segments than the grammar allows", path)
		}
		if profile, err = checkIdent("profile", seg[1], profilePattern); err != nil {
			return "", "", err
		}
		if len(seg) == 3 {
			if subject, err = checkIdent("subject", seg[2], subjectPattern); err != nil {
				return "", "", err
			}
		}
		return profile, subject, nil
	}
	if len(seg) > 1 {
		return "", "", malformedf("path %q is not \"/w/<profile>\", \"/<subject>\", or both", path)
	}
	if subject, err = checkIdent("subject", seg[0], subjectPattern); err != nil {
		return "", "", err
	}
	return "", subject, nil
}

// checkIdent rejects the traversal names before the pattern. Both patterns
// already exclude them by requiring a leading alphanumeric; the explicit
// check states the invariant so a later pattern edit cannot quietly reopen
// the one input class here that could escape a directory.
func checkIdent(what, s string, pattern *regexp.Regexp) (string, error) {
	if s == "." || s == ".." {
		return "", malformedf("%s %q is a directory reference, not a name", what, s)
	}
	if !pattern.MatchString(s) {
		return "", malformedf("%s %q is not a plausible %s", what, s, what)
	}
	return s, nil
}

// validateQuery enforces the size limits on the raw query. It never inspects
// what a parameter means — there is no key allowlist; see the package
// comment. A parameter without "=" counts as one with an empty value, the
// same read the UI's URLSearchParams gives it.
//
// An empty query is legal here. Whether an action can act without one is that
// action's rule: ActionView cannot, and its handler says so.
func validateQuery(q string) error {
	if q == "" {
		return nil
	}
	// net/url already rejects control bytes anywhere in the URL (measured),
	// but this package's contract should not rest on the parser's internals:
	// the scan stays so a parser swap cannot reopen it.
	for i := 0; i < len(q); i++ {
		if q[i] < 0x20 || q[i] == 0x7f {
			return malformedf("control byte 0x%02x in query", q[i])
		}
	}
	parts := strings.Split(q, "&")
	if len(parts) > maxParams {
		return malformedf("%d parameters, over the limit of %d", len(parts), maxParams)
	}
	seen := make(map[string]struct{}, len(parts))
	for _, part := range parts {
		key, value, _ := strings.Cut(part, "=")
		if !keyPattern.MatchString(key) {
			return malformedf("parameter key %q does not match the key shape", key)
		}
		if len(value) > maxValueLen {
			return malformedf("value for %q is %d bytes, over the %d-byte limit", key, len(value), maxValueLen)
		}
		// Two values for one axis is ambiguous, and every producer in this
		// repo emits each key once — reject rather than pick.
		if _, dup := seen[key]; dup {
			return malformedf("duplicate parameter key %q", key)
		}
		seen[key] = struct{}{}
	}
	return nil
}

func malformedf(format string, args ...any) error {
	return fmt.Errorf("%w: %s", ErrMalformed, fmt.Sprintf(format, args...))
}

File: mobile/src/lib/deeplink.ts
/*
 * GDK-873 — `gadak://` links, from iOS to this app.
 *
 * An agent, the desktop app, or a chat message hands over a link
 * (`gadak views open --json` prints one as `deep_link`, cmd/gadak/views.go)
 * and one tap opens that issue here. This module is the pure half: text in,
 * a decision out. Nothing here touches the store, the network, or Tauri —
 * see entry.ts for the wiring, which is what makes this testable without a
 * simulator.
 *
 * ── The grammar is NOT owned here ──
 *
 * internal/deeplink/deeplink.go owns it. This is a second implementation of
 * the same rules because the phone cannot run Go, and a second
 * implementation of a grammar is precisely the drift this repo has already
 * paid for once. So the two are pinned to one table:
 * internal/deeplink/testdata/vectors.json is emitted by the Go parser and
 * consumed, case for case, by deeplink.test.ts. Changing the grammar means
 * regenerating that file, which turns this suite red until parseGadakUrl
 * agrees.
 *
 * The parse is hand-written rather than built on `new URL()` on purpose.
 * Measured while writing this: WHATWG URL and Go's net/url disagree on two
 * inputs that matter here — WHATWG normalizes `/w/../x` down to `/x`, so a
 * traversal Go refuses would have been accepted, and it splits `view:7777`
 * into host+port, so a link with a port would have parsed as the `view`
 * action instead of being refused. Both are exactly the class of input a
 * hostile page would try.
 *
 * ── Security posture (inherited from the Go package comment) ──
 *
 * Any web page can embed a gadak:// link, so the worst one may achieve is
 * that the user briefly looks at the wrong thing. The scheme carries no
 * verb: a link says where to go, never what to do. Nothing in this file may
 * write, submit, pair, or dial.
 *
 * The specific risk this module refuses (GDK-873): a link carrying
 * `/w/<profile>` names a workspace mount on someone's *desktop*. The phone
 * has no profile concept at all — it holds one paired host and the issues
 * that host serves (there is no `/w/` anywhere in mobile/src) — so a
 * profile-bearing link cannot be verified here and must not be treated as
 * this phone's. Opening `GDK-119` from a stranger's workspace on this
 * phone's mirror would show a different issue with the same key, silently.
 */

/** The scheme, matching internal/deeplink's `Scheme` constant. */
export const SCHEME = 'gadak'

/* ── grammar limits, mirroring internal/deeplink/deeplink.go ── */

const MAX_INPUT_LEN = 2048
const MAX_PARAMS = 32
const MAX_VALUE_LEN = 512
const MAX_KEY_LEN = 64

const ACTION_PATTERN = /^[a-z][a-z0-9-]{0,31}$/
const PROFILE_PATTERN = /^[A-Za-z0-9][A-Za-z0-9_.-]{0,63}$/
const SUBJECT_PATTERN = /^[A-Za-z0-9][A-Za-z0-9_.@+-]{0,127}$/
const KEY_PATTERN = new RegExp(`^[a-z][a-z0-9_.]{0,${MAX_KEY_LEN - 1}}$`)

/** The path segment that introduces a profile — the web UI's /w/<name>. */
const PROFILE_SEGMENT = 'w'

/** The only action the Go parser knows by name, and the one agents emit. */
export const ACTION_VIEW = 'view'

/** A parsed link. Field for field the Go `Link` struct. */
export interface GadakLink {
  /** Lowercased. An unrecognised one is a link from a newer gadak. */
  action: string
  /** '' means the primary mirror. */
  profile: string
  /** '' when the action needs none. */
  subject: string
  /** The raw query, verbatim: never decoded, re-encoded, or re-ordered. */
  hash: string
}

/**
 * Why a link was refused.
 *
 * `not_gadak` and `malformed` are the Go parser's two error classes and are
 * kept distinct for the same reason: another app's scheme reaching a shared
 * handler must pass in silence, while a link that IS ours and was refused
 * should say so.
 */
export type ParseFailure = 'not_gadak' | 'malformed'

export type ParseResult =
  | { ok: true; link: GadakLink }
  | { ok: false; failure: ParseFailure; why: string }

function no(failure: ParseFailure, why: string): ParseResult {
  return { ok: false, failure, why }
}

/**
 * Parses a `gadak://` URL. Pure: no I/O, no store, no globals.
 *
 * Returns `not_gadak` for a URL that is not ours (stay silent) and
 * `malformed` for one that is ours and refused (say so).
 */
export function parseGadakUrl(raw: string): ParseResult {
  const prefix = `${SCHEME}:`
  if (raw.slice(0, prefix.length).toLowerCase() !== prefix) {
    return no('not_gadak', 'not a gadak:// URL')
  }
  if (raw.length > MAX_INPUT_LEN) {
    return no('malformed', `input is ${raw.length} bytes, over the ${MAX_INPUT_LEN}-byte limit`)
  }
  // The grammar carries no fragment. Checked on the raw string so a bare
  // trailing '#' is caught too — the Go parser makes the same note.
  if (raw.includes('#')) return no('malformed', 'fragment not allowed')

  let rest = raw.slice(prefix.length)
  if (!rest.startsWith('//')) {
    // The opaque form (`gadak:view?…`) leaves the action empty in Go, which
    // then fails the action pattern. Same outcome, said plainly.
    return no('malformed', 'action "" is not a plain lowercase word')
  }
  rest = rest.slice(2)

  // The authority runs to the first '/' or '?'.
  let end = rest.length
  for (let i = 0; i < rest.length; i++) {
    const c = rest[i]
    if (c === '/' || c === '?') {
      end = i
      break
    }
  }
  const authority = rest.slice(0, end)
  const after = rest.slice(end)

  if (authority.includes('@')) return no('malformed', 'userinfo not allowed')
  // The port stays in the authority (Go keeps it in u.Host), so "view:7777"
  // fails the pattern rather than parsing as the "view" action.
  const action = authority.toLowerCase()
  if (!ACTION_PATTERN.test(action)) {
    return no('malformed', `action ${JSON.stringify(authority)} is not a plain lowercase word`)
  }

  const q = after.indexOf('?')
  const rawPath = q < 0 ? after : after.slice(0, q)
  const query = q < 0 ? '' : after.slice(q + 1)

  // Go validates the DECODED path, so an encoded separator can only add
  // segments — which the segment count and per-segment patterns then reject.
  let path: string
  try {
    path = decodeURIComponent(rawPath)
  } catch {
    return no('malformed', 'unparseable URL: bad percent-encoding in the path')
  }

  const split = splitPath(path)
  if ('why' in split) return no('malformed', split.why)

  const badQuery = validateQuery(query)
  if (badQuery) return no('malformed', badQuery)

  return { ok: true, link: { action, profile: split.profile, subject: split.subject, hash: query } }
}

/**
 * Reads the optional profile and subject out of the decoded path.
 *
 * Accepted: '', '/', '/w/<profile>', '/<subject>', '/w/<profile>/<subject>',
 * each with an optional trailing slash. A leading 'w' segment is reserved as
 * the profile introducer, so a subject may not be the single letter 'w'.
 */
function splitPath(path: string): { profile: string; subject: string } | { why: string } {
  const trimmed = path.replace(/^\/+/, '').replace(/\/+$/, '')
  if (trimmed === '') return { profile: '', subject: '' }
  const seg = trimmed.split('/')
  if (seg[0] === PROFILE_SEGMENT) {
    if (seg.length < 2) return { why: `empty profile after /${PROFILE_SEGMENT}/` }
    if (seg.length > 3) return { why: `path ${JSON.stringify(path)} has more segments than the grammar allows` }
    const profile = checkIdent('profile', seg[1]!, PROFILE_PATTERN)
    if ('why' in profile) return profile
    if (seg.length === 3) {
      const subject = checkIdent('subject', seg[2]!, SUBJECT_PATTERN)
      if ('why' in subject) return subject
      return { profile: profile.value, subject: subject.value }
    }
    return { profile: profile.value, subject: '' }
  }
  if (seg.length > 1) {
    return { why: `path ${JSON.stringify(path)} is not "/w/<profile>", "/<subject>", or both` }
  }
  const subject = checkIdent('subject', seg[0]!, SUBJECT_PATTERN)
  if ('why' in subject) return subject
  return { profile: '', subject: subject.value }
}

/**
 * Rejects the traversal names before the pattern. Both patterns already
 * exclude them by requiring a leading alphanumeric; the explicit check
 * states the invariant so a later pattern edit cannot quietly reopen it.
 */
function checkIdent(what: string, s: string, pattern: RegExp): { value: string } | { why: string } {
  if (s === '.' || s === '..') {
    return { why: `${what} ${JSON.stringify(s)} is a directory reference, not a name` }
  }
  if (!pattern.test(s)) return { why: `${what} ${JSON.stringify(s)} is not a plausible ${what}` }
  return { value: s }
}

/**
 * Enforces the size limits on the raw query. It never inspects what a
 * parameter means — there is deliberately no key allowlist, because the keys
 * the UI understands are owned by web/src/lib/view-config and a second copy
 * would drift silently in the direction that matters.
 */
function validateQuery(q: string): string | null {
  if (q === '') return null
  for (let i = 0; i < q.length; i++) {
    const c = q.charCodeAt(i)
    if (c < 0x20 || c === 0x7f) {
      return `control byte 0x${c.toString(16).padStart(2, '0')} in query`
    }
  }
  const parts = q.split('&')
  if (parts.length > MAX_PARAMS) {
    return `${parts.length} parameters, over the limit of ${MAX_PARAMS}`
  }
  const seen = new Set<string>()
  for (const part of parts) {
    const eq = part.indexOf('=')
    const key = eq < 0 ? part : part.slice(0, eq)
    const value = eq < 0 ? '' : part.slice(eq + 1)
    if (!KEY_PATTERN.test(key)) return `parameter key ${JSON.stringify(key)} does not match the key shape`
    if (value.length > MAX_VALUE_LEN) {
      return `value for ${JSON.stringify(key)} is ${value.length} bytes, over the ${MAX_VALUE_LEN}-byte limit`
    }
    if (seen.has(key)) return `duplicate parameter key ${JSON.stringify(key)}`
    seen.add(key)
  }
  return null
}

/* ── what the phone does with a parsed link ── */

/**
 * What a link asks this app to show. `open_issue` is the only target today,
 * which is a fact about the phone's screens rather than a limit of the
 * scheme: the phone has no router, only `app.detail` and a tab
 * (lib/store.svelte.ts), so an issue is the one thing that can be addressed.
 */
export type DeepLinkTarget = { kind: 'open_issue'; key: string }

/**
 * Why a well-formed link was not acted on.
 *
 * `unsupported_action` is deliberately distinct from a malformed link, the
 * same split desktop/deeplink.go makes: the grammar is fixed but the set of
 * actions grows, so this is what an older build says when it meets a newer
 * link — an upgrade prompt, not "your link is wrong".
 */
export type ResolveRefusal =
  | 'unsupported_action'
  | 'other_workspace'
  | 'no_subject_expected'
  | 'no_target'

export type ResolveResult =
  | { ok: true; target: DeepLinkTarget }
  | { ok: false; refusal: ResolveRefusal; why: string }

/**
 * The issue key inside a view hash. Matches the desk's `issue` place param
 * (web/src/lib/url-state.ts registers it), and the value is checked against
 * the shape a tracker key has so a hash param cannot smuggle a path.
 */
const ISSUE_KEY_PATTERN = /^[A-Za-z][A-Za-z0-9_]{0,31}-[0-9]{1,10}$/

/**
 * Turns a parsed link into what this app should show.
 *
 * Only `view` is honoured, and only its `issue=` param — that is what
 * `gadak views open` emits and what the phone can actually display. A view
 * link with filters but no issue is well-formed and simply has no phone
 * target: the phone has no filtered-list URL to land on.
 */
export function resolveDeepLink(link: GadakLink): ResolveResult {
  if (link.action !== ACTION_VIEW) {
    return { ok: false, refusal: 'unsupported_action', why: `this link needs a newer gadak (action "${link.action}")` }
  }
  if (link.profile !== '') {
    // The refusal this issue exists to make. See the module comment.
    return {
      ok: false,
      refusal: 'other_workspace',
      why: `this link is for the "${link.profile}" workspace on a computer; this phone is paired to one host and has no workspaces`,
    }
  }
  if (link.subject !== '') {
    return { ok: false, refusal: 'no_subject_expected', why: `a view link takes no subject, got "${link.subject}"` }
  }
  const key = issueKeyFromHash(link.hash)
  if (!key) {
    return { ok: false, refusal: 'no_target', why: 'this link points at a filtered list, which the phone cannot open' }
  }
  return { ok: true, target: { kind: 'open_issue', key } }
}

/**
 * Reads `issue=<KEY>` out of a view hash. Uses URLSearchParams for the
 * decoding only — the hash has already passed the grammar's shape and size
 * checks above, so this is not a second parser of the link.
 */
export function issueKeyFromHash(hash: string): string | null {
  if (hash === '') return null
  const raw = new URLSearchParams(hash).get('issue')
  if (!raw) return null
  const key = raw.trim()
  return ISSUE_KEY_PATTERN.test(key) ? key : null
}

/**
 * The whole decision in one call: text from the OS, an action or a reason.
 * Exported as the single entry point so the wiring has nothing to decide.
 */
export function decideDeepLink(
  raw: string,
): { ok: true; target: DeepLinkTarget } | { ok: false; reason: ParseFailure | ResolveRefusal; why: string } {
  const parsed = parseGadakUrl(raw)
  if (!parsed.ok) return { ok: false, reason: parsed.failure, why: parsed.why }
  const resolved = resolveDeepLink(parsed.link)
  if (!resolved.ok) return { ok: false, reason: resolved.refusal, why: resolved.why }
  return { ok: true, target: resolved.target }
}
